File: centrallix/htmlgen/htdrv_image.c
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include "ht_render.h"
#include "obj.h"
#include "cxlib/mtask.h"
#include "cxlib/xarray.h"
#include "cxlib/xhash.h"
#include "cxlib/mtsession.h"
#include "cxlib/strtcpy.h"

/************************************************************************/
/* Centrallix Application Server System 				*/
/* Centrallix Core       						*/
/* 									*/
/* Copyright (C) 1998-2026 LightSys Technology Services, Inc.		*/
/* 									*/
/* This program is free software; you can redistribute it and/or modify	*/
/* it under the terms of the GNU General Public License as published by	*/
/* the Free Software Foundation; either version 2 of the License, or	*/
/* (at your option) any later version.					*/
/* 									*/
/* This program is distributed in the hope that it will be useful,	*/
/* but WITHOUT ANY WARRANTY; without even the implied warranty of	*/
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the	*/
/* GNU General Public License for more details.				*/
/* 									*/
/* You should have received a copy of the GNU General Public License	*/
/* along with this program; if not, write to the Free Software		*/
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  		*/
/* 02111-1307  USA							*/
/*									*/
/* A copy of the GNU General Public License has been included in this	*/
/* distribution in the file "COPYING".					*/
/* 									*/
/* Module: 	htdrv_image.c						*/
/* Author:	Greg Beeley (GRB)					*/
/* Creation:	February 5, 2004 					*/
/* Description:	HTML Widget driver for an image.           		*/
/************************************************************************/


/** globals **/
static struct 
    {
    int		idcnt;
    }
    HTIMG;


/*** htimgSetup - do overall setup work to cover all image widgets.
 ***/
int
htimgSetup(pHtSession s)
    {

	/** Global style code for all image widget "img" tags **/
	htrAddStylesheetItem(s,
	    "\t\timg.wimage { "
		"display:block; "
		"position:relative; "
		"left:0px; "
		"top:0px; "
	    "}\n"
	);

	/** Global style code for all image widget "div" containers **/
	htrAddStylesheetItem(s,
	    "\t\tdiv.wimage { "
		"visibility:inherit; "
		"position:absolute; "
		"overflow:hidden; "
	    "}\n"
	);

    return 0;
    }


/*** htimgRender - generate the HTML code for the label widget.
 ***/
int
htimgRender(pHtSession s, pWgtrNode tree, int z)
    {
    char* ptr;
    char name[64];
    char src[256];
    int x=-1,y=-1,w,h;
    int rval = -1;
    char fieldname[HT_FIELDNAME_SIZE];
    char form[64];
    char* alt_text = NULL;
    char* aspect = NULL;
    char* default_alt_text = "";
    char* default_aspect = "stretch";

	/** Get an id for this. **/
	const int id = (HTIMG.idcnt++);

	/** Verify browser capabilities. **/
	if (!s->Capabilities.Dom1HTML || !s->Capabilities.Dom2CSS)
	    {
	    mssError(1, "HTIMG", "Unsupported browser: W3C DOM1 HTML and DOM2 CSS support required.");
	    goto end;
	    }

    	/** Get x,y,w,h of this object **/
	if (wgtrGetPropertyValue(tree,"x",DATA_T_INTEGER,POD(&x)) != 0) x=0;
	if (wgtrGetPropertyValue(tree,"y",DATA_T_INTEGER,POD(&y)) != 0) y=0;
	if (wgtrGetPropertyValue(tree,"width",DATA_T_INTEGER,POD(&w)) != 0) 
	    {
	    mssError(1,"HTIMG","Image widget must have a 'width' property");
	    goto end;
	    }
	if (wgtrGetPropertyValue(tree,"height",DATA_T_INTEGER,POD(&h)) != 0)
	    {
	    mssError(1,"HTIMG","Image widget must have a 'height' property");
	    goto end;
	    }

	if(wgtrGetPropertyValue(tree,"text",DATA_T_STRING,POD(&ptr)) == 0)
	    alt_text=nmSysStrdup(ptr);
	if (alt_text == NULL) alt_text = default_alt_text;

	/** Image aspect scaling: stretch or preserve **/
	if(wgtrGetPropertyValue(tree,"aspect",DATA_T_STRING,POD(&ptr)) == 0)
	    aspect=nmSysStrdup(ptr);
	if (aspect == NULL) aspect = default_aspect;

	/** Get name **/
	if (wgtrGetPropertyValue(tree,"name",DATA_T_STRING,POD(&ptr)) != 0) goto end;
	strtcpy(name,ptr,sizeof(name));

	/** image source **/
	ptr = "";
	htrCheckAddExpression(s, tree, name, "source");
	wgtrGetPropertyValue(tree,"source",DATA_T_STRING,POD(&ptr));
	strtcpy(src, ptr, sizeof(src));

	htrCheckAddExpression(s, tree, name, "scale");
	htrCheckAddExpression(s, tree, name, "xoffset");
	htrCheckAddExpression(s, tree, name, "yoffset");

	/** Field name **/
	if (wgtrGetPropertyValue(tree,"fieldname",DATA_T_STRING,POD(&ptr)) == 0)
	    strtcpy(fieldname,ptr,sizeof(fieldname));
	else
	    fieldname[0]='\0';
	if (wgtrGetPropertyValue(tree,"form",DATA_T_STRING,POD(&ptr)) == 0)
	    strtcpy(form,ptr,sizeof(form));
	else
	    form[0]='\0';

 	/** Link the widget to the DOM node. **/
	if (htrAddWgtrObjLinkage_va(s, tree, "img%POS", id) != 0) goto end;
	
	/** Initialize image scripts. **/
	if (htrAddScriptInclude(s, "/sys/js/htdrv_image.js", 0) != 0) goto end;
	if (htrAddScriptInit_va(s,
	    "\tim_init(wgtrGetNodeRef(ns, '%STR&SYM'), { "
		"field:'%STR&JSSTR', "
		"form:'%STR&JSSTR', "
	    "});\n", 
	    name,
	    fieldname, form
	) != 0)
	    {
	    mssError(0, "HTDT", "Failed to write JS init call.");
	    goto end;
	    }

	/** Event Handlers **/
	if (htrAddEventHandlerFunction(s, "document", "MOUSEDOWN", "img", "im_mousedown") != 0) goto end;
	if (htrAddEventHandlerFunction(s, "document", "MOUSEMOVE", "img", "im_mousemove") != 0) goto end;
	if (htrAddEventHandlerFunction(s, "document", "MOUSEOUT",  "img", "im_mouseout") != 0) goto end;
	if (htrAddEventHandlerFunction(s, "document", "MOUSEOVER", "img", "im_mouseover") != 0) goto end;
	if (htrAddEventHandlerFunction(s, "document", "MOUSEUP",   "img", "im_mouseup") != 0) goto end;
	
	/** Write the style for the image container div. **/
	if (htrAddStylesheetItem_va(s,
	    "\t\t#img%POS { "
		"left:"ht_flex_format"; "
		"top:"ht_flex_format"; "
		"width:"ht_flex_format"; "
		"height:"ht_flex_format"; "
		"z-index:%POS; "
		"text-align:center; "
	    "}\n",
	    id,
	    ht_flex_x(x, tree),
	    ht_flex_y(y, tree),
	    ht_flex_w(w, tree),
	    ht_flex_h(h, tree),
	    z
	) != 0)
	    {
	    mssError(0, "HTDT", "Failed to write image CSS.");
	    goto end;
	    }

	/** Use a style that honors the aspect ratio attribute. **/
	char* style = (strcmp(aspect, "stretch") == 0)
	    ? /* "stretch" */
	    "width:100%; "
	    "height:100%; "
	    : /* "preserve" */
	    "width:100%; "
	    "height:auto; "
	    "max-width:fit-content; "
	    "max-height:fit-content; "
	    "display:inline; ";
	
	/** Write image HTML, including the containing div. **/
	if (htrAddBodyItemLayer_va(s, 0,
	    "img%POS", id, "wimage",
	    "\n<img "
		"class='wimage' "
		"id='im%POS' "
		"width='%POS' "
		"height='%POS' "
		"style='%STR' "
		"src='%STR&HTE' "
		"alt='%STR&HTE' "
	    ">\n",
	    id,
	    w,
	    h,
	    style,
	    src,
	    alt_text
	) != 0)
	    {
	    mssError(0, "HTDT", "Failed to write image HTML.");
	    goto end;
	    }

	/** Render children. **/
	if (htrRenderSubwidgets(s, tree, z + 1) != 0) goto end;

	/** Success. **/
	rval = 0;

    end:
	if (rval != 0)
	    {
	    mssError(0, "HTIMG",
		"Failed to render \"%s\":\"%s\" (id: %d).",
		tree->Name, tree->Type, id
	    );
	    }
	
	/** Clean up. **/
	if (alt_text != NULL && alt_text != default_alt_text) nmSysFree(alt_text);
	if (aspect != NULL && aspect != default_aspect) nmSysFree(aspect);
	
	return rval;
    }


/*** htimgInitialize - register with the ht_render module.
 ***/
int
htimgInitialize()
    {
    pHtDriver drv;

    	/** Allocate the driver **/
	drv = htrAllocDriver();
	if (!drv) return -1;

	/** Fill in the structure. **/
	strcpy(drv->Name,"DHTML Image Widget");
	strcpy(drv->WidgetName,"image");
	drv->Setup = htimgSetup;
	drv->Render = htimgRender;

	/** Events **/ 
	htrAddEvent(drv,"Click");
	htrAddEvent(drv,"MouseUp");
	htrAddEvent(drv,"MouseDown");
	htrAddEvent(drv,"MouseOver");
	htrAddEvent(drv,"MouseOut");
	htrAddEvent(drv,"MouseMove");

	/** Register. **/
	htrRegisterDriver(drv);

	htrAddSupport(drv, "dhtml");

	HTIMG.idcnt = 0;

    return 0;
    }

File: centrallix-os/sys/js/htdrv_image.js
// htdrv_image.js
//
// Copyright (C) 2004 LightSys Technology Services, Inc.
//
// You may use these files and this library under the terms of the
// GNU Lesser General Public License, Version 2.1, contained in the
// included file "COPYING" or http://www.gnu.org/licenses/lgpl.txt.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.

function im_mouseup(e)
    {
    if (e.kind == 'im') 
	{
	cn_activate(e.layer, 'Click');
	cn_activate(e.layer, 'MouseUp');
	}
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function im_mousedown(e)
    {
    if (e.kind == 'im') cn_activate(e.layer, 'MouseDown');
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function im_mouseover(e)
    {
    if (e.kind == 'im') cn_activate(e.layer, 'MouseOver');
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function im_mouseout(e)
    {
    if (e.kind == 'im') cn_activate(e.layer, 'MouseOut');
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function im_mousemove(e)
    {
    if (e.kind == 'im') cn_activate(e.layer, 'MouseMove');
    return EVENT_CONTINUE | EVENT_ALLOW_DEFAULT_ACTION;
    }

function im_get_source(a)
    {
    return this.source;
    }

function im_set_source(a, v)
    {
    this.source = v;
    pg_set(this.img, "src", v);
    }

function im_action_load_image(aparam)
    {
    var newurl = '';
    if (typeof aparam.Source != 'undefined')
	newurl = aparam.Source;
    else
	return;

    // parameters
    for(var p in aparam)
	{
	if (p == '_Origin' || p == 'Source' || p == '_EventName') continue;
	var v = aparam[p];
	if (newurl.lastIndexOf('?') > newurl.lastIndexOf('/'))
	    newurl += '&';
	else
	    newurl += '?';
	newurl += (htutil_escape(p) + '=' + htutil_escape(v));
	}

    // session linkage
    if (newurl.substr(0,1) == '/')
	{
	if (newurl.lastIndexOf('?') > newurl.lastIndexOf('/'))
	    newurl += '&';
	else
	    newurl += '?';
	if (aparam.LinkApp !== null && (aparam.LinkApp == 'yes' || aparam.LinkApp == 1))
	    newurl += "cx__akey=" + window.akey;
	else
	    newurl += "cx__akey=" + window.akey.substr(0,49);
	}

    this.source = newurl;
    //pg_set(this.img, "src", newurl);
    pg_serialized_load(this.img, newurl, im_loaded, true);
    }

function im_loaded()
    {
    var l = this.mainlayer;
    }

function im_get_x(a)
    {
    return this.xoffset;
    }

function im_set_x(a, v)
    {
    if (isNaN(v) || v === undefined || v === null)
	this.xoffset = 0;
    else
	this.xoffset = v;
    $(this.img).css({"left":parseInt(this.xoffset) + "px"});
    }

function im_get_y(a)
    {
    return this.yoffset;
    }

function im_set_y(a, v)
    {
    if (isNaN(v) || v === undefined || v === null)
	this.yoffset = 0;
    else
	this.yoffset = v;
    $(this.img).css({"top":parseInt(this.yoffset) + "px"});
    }

function im_get_scale(a)
    {
    return this.scale;
    }

function im_set_scale(a, v)
    {
    if (isNaN(v) || v === undefined || v === null)
	this.scale = 1.0;
    else
	this.scale = v;
    $(this.img).width(parseInt(this.scale * $(this).width()));
    }

function im_action_offset(aparam)
    {
    im_set_x.call(this, "xoffset", aparam.X);
    im_set_y.call(this, "yoffset", aparam.Y);
    }

function im_action_scale(aparam)
    {
    im_set_scale.call(this, "scale", aparam.Scale);
    }

function im_init(l)
    {
    htr_init_layer(l,l,"im");
    ifc_init_widget(l);
    htutil_tag_images(l, "im", l, l);
    var imgs = pg_images(l);
    l.img = imgs[0];
    l.img.onload = im_loaded;
    l.source = pg_get(l.img, "src");

    // Image display controls
    l.scale = 1.0;
    l.xoffset = 0;
    l.yoffset = 0;

    // Events
    var ie = l.ifcProbeAdd(ifEvent);
    ie.Add("Click");
    ie.Add("MouseDown");
    ie.Add("MouseUp");
    ie.Add("MouseOver");
    ie.Add("MouseOut");
    ie.Add("MouseMove");

    var iv = l.ifcProbeAdd(ifValue);
    iv.Add("source", im_get_source, im_set_source);
    iv.Add("xoffset", im_get_x, im_set_x);
    iv.Add("yoffset", im_get_y, im_set_y);
    iv.Add("scale", im_get_scale, im_set_scale);

    var ia = l.ifcProbeAdd(ifAction);
    ia.Add("LoadImage", im_action_load_image);
    ia.Add("Offset", im_action_offset);
    ia.Add("Scale", im_action_scale);

    im_loaded.call(this);

    return l;
    }


// Load indication
if (window.pg_scripts) pg_scripts['htdrv_image.js'] = true;
